File: app/src/main/java/com/solutions/alphil/zambiajobalerts/classes/JobsAdapter.java
package com.solutions.alphil.zambiajobalerts.classes;

import android.app.AlertDialog;
import android.content.ActivityNotFoundException;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.google.android.gms.ads.nativead.AdChoicesView;
import com.google.android.gms.ads.nativead.NativeAd;
import com.google.android.gms.ads.nativead.NativeAdView;
import com.solutions.alphil.zambiajobalerts.R;

import java.util.List;

public class JobsAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private List<Object> displayItems;
    private OnJobClickListener listener;

    private static final int VIEW_TYPE_JOB = 0;
    private static final int VIEW_TYPE_AD = 1;


    public interface OnJobClickListener {
        void onJobClick(Job job);
        default void onGenerateCv(Job job) {}
        default void onGenerateCoverLetter(Job job) {}
    }

    public JobsAdapter(List<Object> displayItems, OnJobClickListener listener) {
        this.displayItems = displayItems;
        this.listener = listener;
    }

    @Override
    public int getItemViewType(int position) {
        Object item = displayItems.get(position);
        return item instanceof NativeAd ? VIEW_TYPE_AD : VIEW_TYPE_JOB;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == VIEW_TYPE_AD) {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_native_ad, parent, false);
            return new NativeAdViewHolder(view);
        } else {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_job, parent, false);
            return new JobViewHolder(view);
        }
    }
    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        Object item = displayItems.get(position);
        if (holder instanceof JobViewHolder) {
            Job job = (Job) item;
            ((JobViewHolder) holder).bind(job, listener); // Pass listener to bind method
        } else if (holder instanceof NativeAdViewHolder) {
            NativeAd nativeAd = (NativeAd) item;
            ((NativeAdViewHolder) holder).bind(nativeAd);
        }
    }

    @Override
    public int getItemCount() {
        return displayItems.size();
    }

    public void updateDisplayList(List<Object> newItems) {
        this.displayItems = newItems;
        notifyDataSetChanged();
    }

    @Override
    public void onViewRecycled(@NonNull RecyclerView.ViewHolder holder) {
        super.onViewRecycled(holder);
        if (holder instanceof NativeAdViewHolder) {
            NativeAdViewHolder adHolder = (NativeAdViewHolder) holder;
            // Manually clear the views to unbind the ad and prevent crash
            adHolder.adHeadline.setText("");
            adHolder.adBody.setText("");
            adHolder.adAdvertiser.setText("");
            adHolder.adCallToAction.setText("");
            adHolder.adAppIcon.setImageDrawable(null);
            // Do not call setNativeAd(null) - this causes the NPE in SDK v24.6.0
            // Optionally, if you have a reference to the NativeAd in the holder, call nativeAd.destroy() here
        }
        /*if (holder instanceof NativeAdViewHolder) {
            ((NativeAdViewHolder) holder).nativeAdView.setNativeAd(null);
        }*/
    }

    static class JobViewHolder extends RecyclerView.ViewHolder {
        TextView tvTitle, tvExcerpt, tvDate, tvCompany, tvLocation, tvJobType;
        ImageView ivImage;
        Button btnApplyNow, btnShare, btnDetails;
        Button btnGenerateCv;
        Button btnGenerateCoverLetter;

        JobViewHolder(@NonNull View itemView) {
            super(itemView);
            tvTitle = itemView.findViewById(R.id.tvTitle);
            tvExcerpt = itemView.findViewById(R.id.tvExcerpt);
            tvDate = itemView.findViewById(R.id.tvDate);
            tvCompany = itemView.findViewById(R.id.tvCompany);
            tvLocation = itemView.findViewById(R.id.tvLocation);
            tvJobType = itemView.findViewById(R.id.tvJobType);
            ivImage = itemView.findViewById(R.id.ivImage);
            // Initialize buttons
            btnApplyNow = itemView.findViewById(R.id.btnApplyNow);
            btnShare = itemView.findViewById(R.id.btnShare);
            btnDetails = itemView.findViewById(R.id.btnDetails);
            btnGenerateCv = itemView.findViewById(R.id.btnGenerateCv);
            btnGenerateCoverLetter = itemView.findViewById(R.id.btnGenerateCoverLetter);
        }

        void bind(Job job, OnJobClickListener listener) {
            tvTitle.setText(job.getTitle());
            tvExcerpt.setText(job.getExcerpt());
            tvDate.setText(job.getFormattedDate());

            // Set company if available
            String company = job.getCompany();
            if (!company.isEmpty()) {
                tvCompany.setText("Company: " + company);
                tvCompany.setVisibility(View.VISIBLE);
            } else {
                tvCompany.setVisibility(View.GONE);
            }

            // Set location if available
            String location = job.getLocation();
            if (!location.isEmpty()) {
                tvLocation.setText("Location: " + location);
                tvLocation.setVisibility(View.VISIBLE);
            } else {
                tvLocation.setVisibility(View.GONE);
            }

            // Set job type if available
            String jobType = job.getJobType();
            if (!jobType.isEmpty()) {
                tvJobType.setText("Type: " + jobType);
                tvJobType.setVisibility(View.VISIBLE);
            } else {
                tvJobType.setVisibility(View.GONE);
            }

            if (!job.getFeaturedImage().isEmpty()) {
                Glide.with(itemView.getContext()).load(job.getFeaturedImage()).into(ivImage);
            }

            // Setup Apply Now button
            String application = job.getApplication();
            if (application != null && !application.isEmpty()) {
                btnApplyNow.setVisibility(View.VISIBLE);
                btnApplyNow.setOnClickListener(v -> applyForJob(job));
            } else {
                btnApplyNow.setVisibility(View.GONE);
            }

            // Setup Share button
            btnShare.setOnClickListener(v -> shareJob(job));

            // Setup Details button
            btnDetails.setOnClickListener(v -> {
                if (listener != null) {
                    listener.onJobClick(job);
                }
            });
            btnGenerateCv.setOnClickListener(v -> {
                if (listener != null) {
                    listener.onGenerateCv(job);
                }
            });
            btnGenerateCoverLetter.setOnClickListener(v -> {
                if (listener != null) {
                    listener.onGenerateCoverLetter(job);
                }
            });
            itemView.setOnClickListener(v -> {
                if (listener != null) {
                    listener.onJobClick(job);
                }
            });

        }
        /*
        private void applyForJob(Job job) {

            SharedPreferences prefs;
            prefs = itemView.getContext().getSharedPreferences("ad_prefs", Context.MODE_PRIVATE);
            int viewCount = prefs.getInt("job_views", 0) + 1;
            prefs.edit().putInt("job_views", viewCount).apply();

            String application = job.getApplication();
            if (application != null && !application.isEmpty()) {
                String cleanEmail = application.trim();

                Intent intent = new Intent(Intent.ACTION_SEND);
                intent.setType("text/plain");
                intent.putExtra(Intent.EXTRA_EMAIL, new String[]{cleanEmail});
                intent.putExtra(Intent.EXTRA_SUBJECT, "Job Application: " + job.getTitle());
                intent.putExtra(Intent.EXTRA_TEXT, "I am applying for the " + job.getTitle() + " position.");

                try {
                    itemView.getContext().startActivity(Intent.createChooser(intent, "Apply via:"));
                } catch (ActivityNotFoundException e) {
                    Toast.makeText(itemView.getContext(), "No email app found. Email: " + cleanEmail, Toast.LENGTH_LONG).show();

                    // Copy to clipboard as fallback
                    ClipboardManager clipboard = (ClipboardManager) itemView.getContext().getSystemService(Context.CLIPBOARD_SERVICE);
                    ClipData clip = ClipData.newPlainText("Application Email", cleanEmail);
                    clipboard.setPrimaryClip(clip);
                }
            } else {
                Toast.makeText(itemView.getContext(), "No application email provided", Toast.LENGTH_SHORT).show();
            }
        }
*/
        private void applyForJob(Job job) {
            SharedPreferences prefs;
            prefs = itemView.getContext().getSharedPreferences("ad_prefs", Context.MODE_PRIVATE);
            int viewCount = prefs.getInt("job_views", 0) + 1;
            prefs.edit().putInt("job_views", viewCount).apply();
            String application = job.getApplication();
                if (application != null && !application.isEmpty()) {
                    String cleanApp = application.trim();

                    // Check if it's an email (rough check for @ and domain)
                    if (cleanApp.contains("@") && cleanApp.contains(".")) {
                        // Treat as email: Use ACTION_SENDTO with mailto: to target only email apps
                        Intent intent = new Intent(Intent.ACTION_SENDTO);
                        intent.setData(Uri.parse("mailto:" + cleanApp));
                        intent.putExtra(Intent.EXTRA_SUBJECT, "Job Application: " + job.getTitle());
                        intent.putExtra(Intent.EXTRA_TEXT, createEmailBody(job));

                        // Launch directly (no chooser; only email apps will handle mailto)
                        try {
                            itemView.getContext().startActivity(intent);
                        } catch (ActivityNotFoundException e) {
                            // If no email app is found, offer to copy the email
                            offerEmailCopy(cleanApp);
                        }
                    } else if (cleanApp.startsWith("http://") || cleanApp.startsWith("https://")) {
                        // Treat as URL: Use ACTION_VIEW to target only browsers
                        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(cleanApp));
                        try {
                            itemView.getContext().startActivity(intent);
                        } catch (ActivityNotFoundException e) {
                            // Fallback: Offer to copy the URL (adapt offerEmailCopy if needed)
                            offerEmailCopy(cleanApp); // Or create a separate offerUrlCopy method
                        }
                    } else {
                        // Fallback for unrecognized format: Offer to copy
                        offerEmailCopy(cleanApp);
                    }
                }
            }
        private void offerEmailCopy(String email) {
            new AlertDialog.Builder(itemView.getContext())
                    .setTitle("No Email App Found")
                    .setMessage("Would you like to copy the application email address to clipboard?")
                    .setPositiveButton("Copy Email", (dialog, which) -> {
                        ClipboardManager clipboard = (ClipboardManager) itemView.getContext().getSystemService(Context.CLIPBOARD_SERVICE);
                        ClipData clip = ClipData.newPlainText("Job Application Email", email);
                        clipboard.setPrimaryClip(clip);
                        Toast.makeText(itemView.getContext(), "Email copied: " + email, Toast.LENGTH_LONG).show();
                    })
                    .setNegativeButton("Cancel", null)
                    .show();
        }
        private String createEmailBody(Job job) {
            return "Dear Hiring Manager,\n\n" +
                    "I am writing to apply for the position of " + job.getTitle() +
                    " that I found through Zambia Job Alerts.\n\n" +
                    "Please find my application materials attached.\n\n" +
                    "Thank you for your consideration.\n\n" +
                    "Sincerely,\n" +
                    "[Your Name]\n" +
                    "[Your Phone Number]\n" +
                    "[Your Email Address]";
        }
        private void shareJob(Job job) {
            SharedPreferences prefs;
            prefs = itemView.getContext().getSharedPreferences("ad_prefs", Context.MODE_PRIVATE);
            int viewCount = prefs.getInt("job_views", 0) + 1;
            prefs.edit().putInt("job_views", viewCount).apply();

            String playStoreUrl = "https://play.google.com/store/apps/details?id=com.solutions.alphil.zambiajobalerts";
            String download = "Check out Zambia Job Alerts app for the latest job opportunities in Zambia! Download now: " + playStoreUrl;

            String jobLink = generateJobLink(job);
            String shareText = "Check out this job opportunity: " + job.getTitle() +
                    "\n\n" + jobLink +
                    "\n\nVia Zambia Job Alerts App"+ "\n\n" +download;;

            Intent shareIntent = new Intent(Intent.ACTION_SEND);
            shareIntent.setType("text/plain");
            shareIntent.putExtra(Intent.EXTRA_SUBJECT, "Job Opportunity: " + job.getTitle());
            shareIntent.putExtra(Intent.EXTRA_TEXT, shareText);

            itemView.getContext().startActivity(Intent.createChooser(shareIntent, "Share Job via"));
        }

        private String generateJobLink(Job job) {
            // Use the same deep link structure as before
            return job.getLink();//"https://zambiajobalerts.com/job/" + job.getId();
        }

    }

    static class NativeAdViewHolder extends RecyclerView.ViewHolder {
        NativeAdView nativeAdView;
        ImageView adAppIcon;
        TextView adHeadline, adBody, adAdvertiser, adAttribution;  // Add adAttribution
        Button adCallToAction;
        AdChoicesView adChoicesView;  // Fix to AdChoicesView

        NativeAdViewHolder(@NonNull View itemView) {
            super(itemView);
            nativeAdView = itemView.findViewById(R.id.native_ad_view);
            adAppIcon = itemView.findViewById(R.id.ad_app_icon);
            adHeadline = itemView.findViewById(R.id.ad_headline);
            adBody = itemView.findViewById(R.id.ad_body);
            adAdvertiser = itemView.findViewById(R.id.ad_advertiser);
            adCallToAction = itemView.findViewById(R.id.ad_call_to_action);
            adAttribution = itemView.findViewById(R.id.ad_attribution);  // Add this
            adChoicesView = itemView.findViewById(R.id.ad_choices_view);  // Fix ID and type
        }
        void bind(NativeAd nativeAd) {
            // Bind headline (returns String directly)
            String headline = nativeAd.getHeadline();
            if (headline != null) {
                adHeadline.setText(headline);
                adHeadline.setVisibility(View.VISIBLE);
            } else {
                adHeadline.setVisibility(View.INVISIBLE);
            }

            // Bind body (returns String directly)
            String body = nativeAd.getBody();
            if (body != null) {
                adBody.setText(body);
                adBody.setVisibility(View.VISIBLE);
            } else {
                adBody.setVisibility(View.INVISIBLE);
            }

            // Bind call to action (returns String directly)
            String callToAction = nativeAd.getCallToAction();
            if (callToAction != null) {
                adCallToAction.setText(callToAction);
                adCallToAction.setVisibility(View.VISIBLE);
            } else {
                adCallToAction.setVisibility(View.INVISIBLE);
            }

            // Bind icon
            if (nativeAd.getIcon() != null) {
                adAppIcon.setImageDrawable(nativeAd.getIcon().getDrawable());
                adAppIcon.setVisibility(View.VISIBLE);
            } else {
                adAppIcon.setVisibility(View.INVISIBLE);
            }

            // Bind advertiser (returns String directly)
            String advertiser = nativeAd.getAdvertiser();
            if (advertiser != null) {
                adAdvertiser.setText(advertiser);
                adAdvertiser.setVisibility(View.VISIBLE);
            } else {
                adAdvertiser.setVisibility(View.INVISIBLE);
            }


            // Show Ad attribution badge (required for programmatic)
            if (adAttribution != null) {
                adAttribution.setVisibility(View.VISIBLE);
            }

            // Register AdChoicesView (required)
            if (adChoicesView != null) {
                nativeAdView.setAdChoicesView(adChoicesView);
            }

            // Register other views for click tracking
            nativeAdView.setHeadlineView(adHeadline);
            nativeAdView.setBodyView(adBody);
            nativeAdView.setCallToActionView(adCallToAction);
            nativeAdView.setIconView(adAppIcon);
            nativeAdView.setAdvertiserView(adAdvertiser);

            // Set the native ad
            //nativeAdView.setNativeAd(nativeAd);
            // Register the views with the native ad (handles clicks, impressions, etc.)
            nativeAdView.setNativeAd(nativeAd);


        }
    }
}

File: app/src/main/java/com/solutions/alphil/zambiajobalerts/classes/JobsAdapter.kt
package com.solutions.alphil.zambiajobalerts.classes

import android.view.ViewGroup
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Button as ComposeButton
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.ViewCompositionStrategy
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.text.HtmlCompat
import androidx.recyclerview.widget.RecyclerView
import com.google.android.gms.ads.nativead.NativeAd
import com.solutions.alphil.zambiajobalerts.R
import com.solutions.alphil.zambiajobalerts.ui.common.ComposeNativeAd

class JobsAdapter(
    private var displayItems: List<Any>,
    private val listener: OnJobClickListener?,
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    override fun getItemViewType(position: Int): Int =
        if (displayItems[position] is NativeAd) VIEW_TYPE_AD else VIEW_TYPE_JOB

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view = ComposeView(parent.context).apply {
            layoutParams = ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT,
            )
            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnDetachedFromWindow)
        }
        return if (viewType == VIEW_TYPE_AD) NativeAdComposeViewHolder(view) else JobViewHolder(view)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val item = displayItems[position]
        when (holder) {
            is JobViewHolder -> holder.bind(item as Job, listener)
            is NativeAdComposeViewHolder -> holder.bind(item as NativeAd)
        }
    }

    override fun getItemCount(): Int = displayItems.size

    fun updateDisplayList(newItems: List<Any>) {
        displayItems = newItems
        notifyDataSetChanged()
    }

    fun interface OnJobClickListener {
        fun onJobClick(job: Job)
        fun onGenerateCv(job: Job) = Unit
        fun onGenerateCoverLetter(job: Job) = Unit
        fun onToggleSave(job: Job) = Unit
        fun isJobSaved(job: Job): Boolean = false
    }

    private class JobViewHolder(private val composeView: ComposeView) : RecyclerView.ViewHolder(composeView) {

        fun bind(job: Job, listener: OnJobClickListener?) {
            composeView.setContent {
                MaterialTheme {
                    JobListCard(
                        job = job,
                        onApply = { JobActionHandler.applyForJob(composeView.context, job) },
                        onShare = { JobActionHandler.shareJob(composeView.context, job) },
                        onDetails = { listener?.onJobClick(job) },
                        onGenerateCv = { listener?.onGenerateCv(job) },
                        onGenerateCoverLetter = { listener?.onGenerateCoverLetter(job) },
                        isSaved = listener?.isJobSaved(job) == true,
                        onToggleSave = { listener?.onToggleSave(job) },
                    )
                }
            }
        }
    }

    private class NativeAdComposeViewHolder(private val composeView: ComposeView) : RecyclerView.ViewHolder(composeView) {
        fun bind(nativeAd: NativeAd) {
            composeView.setContent {
                MaterialTheme {
                    ComposeNativeAd(nativeAd = nativeAd)
                }
            }
        }
    }

    companion object {
        private const val VIEW_TYPE_JOB = 0
        private const val VIEW_TYPE_AD = 1
    }
}

@Composable
fun JobListCard(
    job: Job,
    onApply: () -> Unit,
    onShare: () -> Unit,
    onDetails: () -> Unit,
    onGenerateCv: () -> Unit,
    onGenerateCoverLetter: () -> Unit,
    isSaved: Boolean = false,
    onToggleSave: () -> Unit = {},
) {
    val primary = Color(0xFF001F3F)
    val orange = Color(0xFFFF851B)
    val hasApplication = job.getApplication().isNotBlank()

    Card(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 8.dp, vertical = 6.dp),
        colors = CardDefaults.cardColors(containerColor = Color.White),
        elevation = CardDefaults.cardElevation(defaultElevation = 3.dp),
    ) {
        Column(modifier = Modifier.padding(16.dp)) {
            Text(
                text = cleanJobText(job.getTitle()),
                color = primary,
                fontSize = 18.sp,
                fontWeight = FontWeight.Bold,
            )

            val company = cleanJobText(job.getCompany())
            val location = cleanJobText(job.getLocation())
            val jobType = cleanJobText(job.getJobType())
            if (company.isNotEmpty()) {
                Text("Company: $company", color = Color(0xFF333333), fontSize = 14.sp)
            }
            if (location.isNotEmpty()) {
                Text("Location: $location", color = Color(0xFF333333), fontSize = 14.sp)
            }
            if (jobType.isNotEmpty()) {
                Text("Type: $jobType", color = Color(0xFF333333), fontSize = 14.sp)
            }

            val excerpt = cleanJobText(job.getExcerpt())
            if (excerpt.isNotEmpty()) {
                Spacer(modifier = Modifier.height(6.dp))
                Text(
                    text = excerpt,
                    color = Color(0xFF333333),
                    fontSize = 14.sp,
                    maxLines = 4,
                    overflow = TextOverflow.Ellipsis,
                )
            }

            val date = cleanJobText(job.getFormattedDate().orEmpty())
            if (date.isNotEmpty()) {
                Spacer(modifier = Modifier.height(4.dp))
                Text(date, color = Color(0xFF666666), fontSize = 12.sp)
            }

            Spacer(modifier = Modifier.height(12.dp))
            Column(verticalArrangement = Arrangement.spacedBy(8.dp), modifier = Modifier.fillMaxWidth()) {
                Row(horizontalArrangement = Arrangement.spacedBy(8.dp), modifier = Modifier.fillMaxWidth()) {
                    ComposeButton(
                        onClick = onApply,
                        modifier = Modifier
                            .weight(1f)
                            .heightIn(min = 48.dp),
                        colors = ButtonDefaults.buttonColors(containerColor = orange),
                        enabled = hasApplication,
                    ) {
                        JobButtonContent(label = "Apply", iconRes = R.drawable.ic_action_apply)
                    }
                    OutlinedButton(
                        onClick = onShare,
                        modifier = Modifier
                            .weight(1f)
                            .heightIn(min = 48.dp),
                    ) {
                        JobButtonContent(label = "Share", iconRes = R.drawable.ic_action_share)
                    }
                    OutlinedButton(
                        onClick = onDetails,
                        modifier = Modifier
                            .weight(1f)
                            .heightIn(min = 48.dp),
                    ) {
                        JobButtonContent(label = "Details", iconRes = R.drawable.ic_action_details)
                    }
                }
                Row(horizontalArrangement = Arrangement.spacedBy(8.dp), modifier = Modifier.fillMaxWidth()) {
                    ComposeButton(
                        onClick = onGenerateCv,
                        modifier = Modifier
                            .weight(1f)
                            .heightIn(min = 48.dp),
                        colors = ButtonDefaults.buttonColors(containerColor = primary),
                    ) {
                        JobButtonContent(label = "CV", iconRes = R.drawable.ic_action_cv)
                    }
                    ComposeButton(
                        onClick = onGenerateCoverLetter,
                        modifier = Modifier
                            .weight(1f)
                            .heightIn(min = 48.dp),
                        colors = ButtonDefaults.buttonColors(containerColor = primary),
                    ) {
                        JobButtonContent(label = "Cover", iconRes = R.drawable.ic_action_cover_letter)
                    }
                    OutlinedButton(
                        onClick = onToggleSave,
                        modifier = Modifier
                            .weight(1f)
                            .heightIn(min = 48.dp),
                    ) {
                        JobButtonContent(
                            label = if (isSaved) "Saved" else "Save",
                            iconRes = if (isSaved) R.drawable.ic_bookmark_filled else R.drawable.ic_bookmark_border,
                        )
                    }
                }
            }
        }
    }
}

@Composable
private fun JobButtonContent(label: String, iconRes: Int) {
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center,
    ) {
        Icon(
            painter = painterResource(iconRes),
            contentDescription = null,
            modifier = Modifier.size(16.dp),
        )
        Spacer(modifier = Modifier.height(2.dp))
        Text(
            text = label,
            fontSize = 11.sp,
            lineHeight = 12.sp,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
            textAlign = TextAlign.Center,
        )
    }
}

private fun cleanJobText(raw: String): String =
    HtmlCompat.fromHtml(raw, HtmlCompat.FROM_HTML_MODE_LEGACY)
        .toString()
        .replace(Regex("\\s+"), " ")
        .trim()
